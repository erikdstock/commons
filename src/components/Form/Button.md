React component example:

```js
<Button>Push Me</Button>
```

```js
<div style={{width: "500px"}}>
  <Button block >Block Button</Button>
</div>
```

```js
  <Button outline>Outline Button</Button>
```


To render an example as highlighted source code add a `static` modifier:

```jsx static
import React from 'react';
```

Examples with all other languages are rendered only as highlighted source code, not an actual component:

```html
<Button size="large">Push Me</Button>
```

Any [Markdown](http://daringfireball.net/projects/markdown/) is **allowed** _here_.
