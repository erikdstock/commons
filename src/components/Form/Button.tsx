import styled from '@emotion/styled'
import { css as ssCss } from "@styled-system/css"
import themeGet from '@styled-system/theme-get'
import React from "react"
import { BoxProps } from "../layout"


const resetButtonStyle = css({
  background: "none",
  color: "inherit",
  border: "none",
  padding: "0",
  font: "inherit",
  cursor: "pointer",
  outline: "inherit",
})

interface ButtonProps extends BoxProps with  {}

// TODO: make disabled button states automatic, rather than variants??
// - could use alpha channel on a button variant's background color
// - could have smart disabled variant`s`, ie
// <Button variant="primary" disabled ...
//    ... => <BaseButton variant={disabled ? "disabled" : variant} ...
//  - This would add requirements to users on their themes though.
export const Button: React.FunctionComponent<ButtonProps> = styled('button')(resetButtonStyle, ssCss({
  bg: themeGet('buttons.primary.bg')
})
