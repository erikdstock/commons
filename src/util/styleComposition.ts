import styled from "@emotion/styled"
import { CSSObject, SystemStyleObject } from "@styled-system/css"
import ssCss from "@styled-system/css"
import get from "lodash/get"
import { RosesTheme } from "../theme"

/** Clumsily force styledCss function to believe that there will be a theme present */
type StyledCssFn = (o: SystemStyleObject) => CSSObject
const styledCss = ssCss as StyledCssFn

/** The base style for Box and everything on up. */
export const boxStyle: CSSObject = {
  boxSizing: "border-box",
}

interface ComposeOptions {
  defaultVariant?: string
}

const themeVariantsRoot: keyof RosesThemeObject = "variants"
const componentStylesRoot: keyof RosesThemeObject = "componentStyles"

interface StyledInterpolationProps extends RosesStyleProps {
  theme: RosesThemeObject
}

type InterpolationFn = (p: StyledInterpolationProps) => CSSObject | void

/** Reach into the theme, get a style object, apply it with styledCss */
const themedComponent: (
  path: string
) => (p: StyledInterpolationProps) => CSSObject | undefined = (
  path: string
) => (props: StyledInterpolationProps) => {
  const componentStyle = get(props.theme, `${componentStylesRoot}.${path}`)
  console.warn("CS: ", componentStyle)
  return componentStyle && styledCss(componentStyle)
}

const rxHandler: (p: StyledInterpolationProps) => CSSObject | undefined = ({
  rx,
}) => rx && styledCss(rx)

type VariantHandler = (
  key: string,
  op: ComposeOptions
) => (p: RosesStyleProps & StyledInterpolationProps) => CSSObject | undefined

const variantHandler: VariantHandler = (componentKey, ops) => ({
  theme,
  variant,
}) => {
  const selectedVariant = variant || ops.defaultVariant
  if (selectedVariant) {
    const base = `${themeVariantsRoot}.${componentKey}`
    const variantCss = get(theme, `${base}.${selectedVariant}`)
    console.warn("variant: ", variantCss)
    return variantCss && styledCss(variantCss)
  }
}

/**
 * Wrap a base styled component with:
 *   - a responsive, theme-aware style `css` prop
 *   - a `variant` prop for accessing styles set on the theme.variants key
 * @example:
 * ```tsx
 *  const RawBox = styled('div')({boxSizing: 'border-box'})
 *  const Widget = withStyleProps('Box', RawBox, {})
 * ```
 */
export const withStyleProps = (
  componentKey: string,
  Component: React.ComponentType<RosesStyleProps>,
  options: ComposeOptions = {} // not using this yet
) => {
  return styled(Component)<RosesStyleProps>(
    themedComponent(componentKey),
    variantHandler(componentKey, options),
    rxHandler
  )
}
// /** Compose in the default style + variants for the key defined in your theme
//  * @example ```
//  * const theme = {
//  *   styles: {
//  *     Button: {
//  *       color: "blue"
//  *     }
//  *   }
//  * }
//  *
//  * const Button = styled('button')(ButtonProps, themed("Button"))
//  * ```
//  */
// export const themed = key => props => props.theme.styles[key]
