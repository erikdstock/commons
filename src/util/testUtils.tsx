import React from "react"
import renderer from "react-test-renderer"
import { RosesTheme } from "../Theme"

export const render = (tree: React.ReactNode) =>
  renderer.create(<RosesTheme>{tree}</RosesTheme>)
