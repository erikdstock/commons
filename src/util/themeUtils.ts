// /** Compose in the default style + variants for the key defined in your theme
//  * @example ```
//  * const theme = {
//  *   styles: {
//  *     Button: {
//  *       color: "blue"
//  *     }
//  *   }
//  * }
//  *
//  * const Button = styled('button')(ButtonProps, themed("Button"))
//  * ```
//  */
// export const themed = key => props => props.theme.styles[key]
