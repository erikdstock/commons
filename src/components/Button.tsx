import styled from "@emotion/styled"
import { withStyleProps } from "../util/styleComposition"

export const Button = withStyleProps({
  name: "Button",
  component: "button",
  defaultVariant: "primary",
})
