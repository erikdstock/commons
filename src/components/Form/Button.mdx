---
Button
---
import { Button } from './Button'
import { Box, Flex } from '../../'
import { Playground } from 'docz'

# Buttons

Standard Rebass Buttons.

### Primary:

<Playground>
  <Flex justifyContent="space-around">
    <Button onClick={() => alert("ðŸŒ¹")}>
      Normie
    </Button>
  </Flex>
</Playground>


### Variants:
<Playground >
  <Flex justifyContent="space-around">
  <Box p={3}>
    <Button variant="outline" onClick={() => alert("ðŸŒˆ")}>Outline</Button>
  </Box>
  <Box p={3}>
    <Button variant="inactive" onClick={() => alert("ðŸ¥€")}>Inactive</Button>
  </Box>
  </Flex>
</Playground>

# Cool



