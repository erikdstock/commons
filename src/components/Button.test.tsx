import React from "react"
import { render } from "../util/testUtils"
import { Button } from "./Button"

describe("Button", () => {
  const tree = render(<Button>Click me</Button>).toJSON()
  it("renders correctly", () => {
    expect(tree).toMatchSnapshot()
  })
})
