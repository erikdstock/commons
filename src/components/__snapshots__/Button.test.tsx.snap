// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders correctly 1`] = `
.emotion-0 {
  display: inline-block;
  background: none;
  font-family: 'Arial',system-ui,sans-serif;
  font-size: 16px;
  color: inherit;
  border: none;
  padding: 8px;
  margin: 4px;
  cursor: pointer;
  outline: inherit;
  border-radius: 5px;
  color: #fff;
  background-color: #07c;
}

.emotion-0:disabled {
  background-color: hsl(0,0%,77%);
  color: #f6f6f6;
}

<button
  className="emotion-0"
>
  Click me
</button>
`;
